readme.md
# LibCraftText

Are you an add-on developer who compares quest names against a list of strings to figure out which quest is a daily crafting writ? Master writ?

This is a library of localized strings for crafting quest recognition, and parser code that uses those strings to convert "Craft Normal Ancestor Silk Shoes" into `{ item=SHOES, material=ANCESTOR_SILK }`

Example dump of tables from parsing some daily crafting writ quests, in EN English, DE German, and JA Japanese:

![/example1 Output Table Dump](doc/img/example1_triptych.jpg)

```Lua
LibCraftText.DailyQuestNameToCraftingType("Enchanter Writ")
---> returns CRAFTING_TYPE_ENCHANTING

LibCraftText.ParseDailyCondition(CRAFTING_TYPE_ENCHANTING
            , "Craft Superb Glyph of Health with Ta: 0/1")
---> returns table:
{
    essence = { name="Oko",    item_id=45831, ... }
    potency = { name="Rejera", item_id=64509, ... }
    aspect  = { name="Ta",     item_id=45850, ... }
}

LibCraftText.ParseDailyCondition(CRAFTING_TYPE_CLOTHIER
            , "Craft Normal Ancestor Silk Shoes: 0/1")
---> returns table:
{
    item     = { name="shoes",         mat_set_id=mat_lgt, pattern_index=3, ... }
    material = { name="Ancestor Silk", mat_set_id=mat_lgt, level_index=10, ... }
}

LibCraftText.ParseDailyCondition(CRAFTING_TYPE_ALCHEMY
            , "Acquire Mudcrab Chitin: 3/3")
---> returns table:
{
    item    = { name = "Mudcrab Chitin", item_id=77591, ... }
}

-- and if your ESO UI is set to DE German or some other supported
-- language, then the name strings change, but the numbers don't:

LibCraftText.ParseDailyCondition(CRAFTING_TYPE_ALCHEMY
            , "Besorgt etwas Schlammkrabbenchitin: 3/3")
---> returns table:
{
    item    = { name = "Schlammkrabbenchitin", item_id=77591, ... }
}
```

## GitHub Repo

GitHub repo at [https://github.com/ziggr/ESO-LibCraftText](https://github.com/ziggr/ESO-LibCraftText).

## Supported Languages

- EN English
- DE German
- FR French
- RU Russian [RuESO](https://www.esoui.com/downloads/info1347-RuESO.html)
- ES Spanish [ESO Spanish Translation](https://www.esoui.com/downloads/info838-ESOSpanishTranslationWIP-74.26.html)
- IT Italian [ESO Italian Translation](https://www.esoui.com/downloads/info1067-ESOItalianTranslationWIP.html) has not yet gotten to crafting writs. Not supported.
- JA Japanese [ESO Japanese Translation](https://www.esoui.com/downloads/info2154-ESO-JapaneseLocalization.html)

---
Begin notes to Future Zig: developing LibCraftText


# Design Goal

The goal is a small collection of public APIs that can take any daily or master crafting writ quest and return a set of enums or constants that tell you what to craft.

This is not a translator: it does not answer "tell me how to say 'Rubedite' in German." (Yes, that very answer lurks buried within the lang_db files. But that's not the point of this add-on library.)

# Development Environment

1. Install translation add-ons. See [Supported Languages](#Supported%20Languages) above for the list.
2. Create 10 alts, one for each crafting rank.
    Spread Provisioning Rank 1-3 across all three alliances. You will eventually need AD DC EP rank 1 for three days, then rank AD DC EP rank 2, then rank 3.
3. Modify `LibCraftText.txt` to enable SavedVariables and LibCraftText_Discover.lua.
    That's how you extract strings from the ESO client.

## Alinor: Best City for Daily Crafting Writs

Alinor is the best city for iterating through daily crafting writs. The daily writ boards are next to each other in the center of the crafting area. Almost no walking required.

Buy the **Alinor Crest Townhouse.** Then you can teleport each new alt to that home, walk out the front door, and you're there. _Much faster_ than level 1 horse rides from Shimmerene.

# Development Cycle

## `/lct discover` Discover item and other names

On your Rank 10 alt with complete trait, reagent, and recipe knowledge:

1. `/lct discover` to scan set, trait, and many, many, other strings.
2. Interact with each of 6 crafting stations: BS CL WW JW EN AL (PR not necessary)
3. `/lct lang` to switch to the next language
4. Repeat 1-3 for each supported language

Your SavedVariables $AccountWide is now filled with lots of text. Absorb it into `lang_db.lua` at your leisure.

Step 2 is enough of a hassle that you really don't want to do it often. Which is why `/lct forget` commented out `LibCraftText.saved_var[field]`. If you *really* want to clear out an hour's worth of station-interacting and language-switching, Uncomment that line.

Discover didn't collect all the text you expected? See `LibCraftText.Discover()`. Most of its steps are commented out to save time. Un-comment-out the parts you need.

## `/lct scanlang` Scan daily quests

Learn daily quest text on each of your 10 alts:

1. Log in to alt N
2. `/lct abandon` to abandon all daily crafting quests and forget them from this alt's saved chariables.
3. Interact with crafting writ boards to acquire 7 daily crafting quests.
4. `/lct scanlang` to scan quest journal, record text, and switch to next language
5. Repeat step 4 until you've looped back to your starting language (usually EN English).
6. (Optional) `/lct scan` to display the line of language codes: make sure they're all green:
 <font face="monospace" color="#669966">EN de fr ru es it ja</font>
7. `/logout`
8. Repeat steps 1-7 for each of 10 alts.

Takes about an hour if you're distracted, or half an hour if you're focused.

Your SavedVariables for each alt is now filled with quest text. Absorb it into `lang_db.lua` at your leisure.

## Absorb SavedVariables into `lang_db.lua`

1. `make getpts`: Copy SavedVariables/LibCraftText.lua into data/ .
2. `make lang`: Absorb SavedVariables into `lang_db.lua` by running `lang_gen.lua`.
    (Also generates lang/xx.lua files, but that's not important right now.)

Have you added code to `LibCraftText_Discover.lua` to collect more text? See ??? below.

## Generate test fodder

Add new text to `gen_test.data.lua` so that we can prevent it from breaking:

1. `make lang`: (Optional) pick up any SavedVariables changes you might have missed and absorb them into `lang_db.lua`.
    (Also generates lang/xx.lua files, but that's not important right now.)
2. `make gentest`: Merge text from `lang_db.lua` into `test/gen_test.data.lua`, inserting new lines for new text.
    Expect test to fail if you've added new lines: new lines without expected results should cause the test to fail until you supply expected values, and parser code to satisfy those expected values.

As of 2018-11-13, the gentest generator, data, and runner all know nothing except for daily crafting writs. They will need to be extended to support master writs.

## Test, Code, Repeat

1. `make lang gentest` to update the generated test data, run it against LibCraftText, and fail or pass.
2. Hand-edit `test/gen_test.data.lua` to supply expected results for new lines.
3. Edit `LibCraftText.lua` parser code, or more likely just the regex arrays above the parser code, until your test passes.

## Bad lang_tables

Occasionally `lang_db.lua` will end up with a bad lang_table: a record that lists one thing in EN English, but a different thing in DE German or whatever. Symptom is a `make gentest` that fails, but when you examine the failure report, you can see that the EN English text does not match the meaning of the DE German text or whatever.

Usually these are daily crafting records, not used in `lang_template.txt`. If so, just delete them.

If they are something more permanent, with a key that appears in `lang_template.txt`, you might need to hand-repair them, or delete, regenerate, and inspect.

# TO DO

- [x] harvest Rolis dialog text
- [x] watch quest status changes as we craft things for a currently active master writ quest (Hey, can Dolgubon's Writ Crafter do this for active quests? That'd be helpful while testing. Don't make me write something...)
- [x] /example2 daily writs works in EN English
- [x] /example2 daily writs works in other languages EN DE FR RU ES JA
- [x] /example3 master writs works in EN English
- [x] /example3 master writs works in other languages EN DE FR RU ES JA
- [x] iterator-generator given a quest_index, visits all steps, all conditions
- [x] public API variants that accept a quest_index and return a flat list of
      parse_result tables, one for each matching condition. Insert into
      parse_result tables additional keys for step_i, step_info, condition_i,
      condition_info, crafting_type.
- [x] Use the above API variants in examples when dumping.
- [x] Rearrange quest/dialog stuff into tables lower in the file
- [x] Find all master writ recipes!
- [x] More master writ test fodder? Meh.
- [x] Use LibCraftText in ZZDailyWrits and WritWorthy_AutoQuest.
- [ ] performance pass: call ParseQuest() in a loop over all quests and see if anything is pointlessly expensive.
- [ ] DRY collapse of all that copypasta

# `/script` commands

## `/script SetCVar("language.2", "en")` switch language

Switches ESO UI to EN English. Reloads UI automatically. Other language codes work.

## `/lct en` switch language
An alias for `SetCVar`. Other language codes work.

## `/lct discover` collect rank-independent text

Iterates through motifs, sets, traits, and more, to collect text for anything that doesn't depend on rank. Takes a long time. Prompts you to interact with each crafting station to gather equipment and other names.

Often much of this `LibCraftText.Discover()` is commented out, because boy howdy it sure takes a long time and hammers the API. Uncomment what you need.

##  `/lct scan` collect rank-dependent daily crafting text

Scans quest journal and collects daily crafting writ strings.

## `/lct lang` switch to next language

## `/lct scanlang` alias for `scan` + `lang`

Scan quest journal, then switch to next language. _Greatly_ accelerates the daily slog through 10 alts.

## `/lct forget` Clear Saved Chariables

Can also clear $AccountWide saved variables, if you comment out one line in `LibCraftText.Forget`

## `/lct abandon` abandon all daily crafting writs

Do this once per alt at the start of each alt's `/lct scanlang` slog.

## `/lct alchemist` learn all alchemy traits

Run this while interacting with an alchemy station to craft 26 potions that learn all 30 alchemy traits on all 28 reagents.

Requires alchemy skill `Laboratory Use` for 3 reagents.

## `/lct rolis` listen for events

(Originally was just for Rolis interaction, now for daily as well but I'm not changing its name.)

Installs event listeners and records quest and dialog strings as they go by. You'll have to manually extract the strings from saved variables and build lang_tables. Google Translate is your friend.


doc/changelog.txt
[B]5.3.1 2020-02-28[/B]
[LIST]
[*] LibStub purge
[/LIST]

[B]5.0.1 2019-05-05[/B]
[LIST]
[*] Version and API bump.
[*] ## IsLibrary: true so that Elsweyr can identify this as a library.
[/LIST]

[B]4.2.5 2018-11-25[/B]
[LIST]
[*] Example2 code for daily writ quests
[*] Example3 code for master writ quests
[*] I finally caved in and added [FONT="Courier New"]for EachCondition(quest_index)[/FONT] utility code for iterating over quest conditions. If I had to write that loop a dozen times, other folks probably will, too.
[*] Also added [FONT="Courier New"]ParseQuest(quest_index)[/FONT] to use that iterator on a quest and return a list of its parsed conditions.
[/LIST]

[B]4.2.4 2018-11-12[/B]
[LIST]
[*] Example1 code
[*] Full parse of all daily crafting conditions, returning results from
[*] Giant tables of crafting items and materials.
[/LIST]

[B]4.2.3 2018-10-27[/B]
[LIST]
[*] ES Spanish
[*] RU Russian
[*] JA Japanese
[*] All daily and master crafting writ names translated.
[*] Still missing Rolis dialog, will deal with later if/when I can get a
    guild hall's tables to actually [i]work[/i] on PTS.
[/LIST]

[B]4.2.2 2018-10-27[/B]
[LIST]
[*] DE German and FR French strings added for daily and master writ
    quest names.
[*] Lots of behind-the-scenes automation to help collect translations from
    ESO client and use them to generate translation files. See GitHub repo
    for details.
[/LIST]

[B]4.2.1 2018-10-26[/B]
[LIST]
[*] Initial release, en-english strings
[/LIST]


data/snark.lua
-- Snarky comments from Rolis when yuo start interacting with them.
--
-- Not all 30 EN lines are translated to all languages.
-- Not all other languages are translations of EN lines.
--
-- You don't need these at all for parsing Rolis dialog. Options are what you
-- need.
--
-- So don't try to match them all up. It's an Amazing Colossal Waste of Time.
--

["snark"] =
{
    ["de"] =
    {
        [1] = "Wenn das nicht der Künstler ist. Mir wurde gesagt, dass gute Arbeit ihre Zeit braucht.",
        [2] = "Wir werden ja schließlich alle nicht jünger, wisst Ihr?",
        [3] = "Ich betrachte das hier gerne als „mittelbares Mäzenatentum“. Also zeigt doch mal her, was Ihr für unseren Mäzen habt.",
        [4] = "Wenn ich sowieso warten muss, könntet Ihr mir wenigstens ein Kissen bringen.",
        [5] = "Je länger die Kunden warten, desto ungeduldiger werden sie. Wenn Ihr wisst, was ich meine.",
        [6] = "Ihr seht aus, als wärt Ihr bereit, ein bisschen Eindruck zu schinden.",
        [7] = "Ich habe alle Zeit der Welt. Der Kunde natürlich nicht.",
        [8] = "Der Künstler trifft ein. Wie läuft die Arbeit?",
        [9] = "Die meisten Kunden warten nicht gerne. So was kann einen Ruf sehr schnell schädigen, wisst Ihr.",
        [10] = "Seid Ihr bereit, die Waren abzugeben?",
        [11] = "Wird auch Zeit. Habt Ihr, was ich brauche?",
        [12] = "Seid Ihr gekommen, um unseren kleinen „Privatkontrakt“ zu erfüllen?",
    },
    ["fr"] =
    {
        [1] = "Ça fera l'affaire. Oui, ça conviendra très bien.",
        [2] = "Vous savez, on ne rajeunit pas, là.",
        [3] = "Content de voir le produit fini. Beau travail.",
        [4] = "Le client sera peut-être satisfait, cette fois-ci.",
        [5] = "Je ferai en sorte que cet objet parvienne à bon port. Voici votre part.",
        [6] = "Excellent. Ceci arrivera entre les bonnes mains.",
        [7] = "Vous savez, là, on ne rajeunit pas.",
        [8] = "J'ai tout le temps du monde. Mais pas le client, bien sûr.",
        [9] = "Quitte à me faire attendre, vous pourriez m'apporter un coussin.",
        [10] = "Il était temps. Vous avez ce qu'il me faut ?",
        [11] = "L'artisan arrive. Ça va, les affaires ?",
        [12] = "La plupart des clients n'aiment pas attendre. C'est mauvais pour la réputation, vous savez.",
        [13] = "Plus un client attend, moins il est patient. Si vous voyez ce que je veux dire.",
        [14] = "Sympa. Comment vous faites pour plier le métal comme ça ? Bref, voilà votre paiement.",
        [15] = "Pas mal du tout. Ça luit bien. Le style est franc. Le client sera content.",
        [16] = "La forme et la fonction, qu'est-ce qui vous importe le plus ? Bah, on s'en cogne. Donnez-moi ça.",
        [17] = "J'aime voir ça comme un « soutien indirect ». Alors, voyons ce que vous avez pour notre soutien.",
        [18] = "Vous venez livrer notre petit « contrat privé » ?",
        [19] = "Tiens, tiens, mon artisan. On dirait que le bon travail, ça prend du temps.",
        [20] = "Vous avez l'air prêt à m'impressionner.",
        [21] = "Vous êtes prêt à livrer les marchandises ?",
    },
    ["ru"] =
    {
        [1] = "Никто из нас не молодеет, знаешь ли.",
        [2] = "Чем дольше клиент ждет, тем меньше терпения у него остается. Если ты понимаешь, о чем я.",
        [3] = "У меня есть все время в мире. А у клиента, конечно, — нет.",
        [4] = "А вот и мастер. Как продвигается работа?",
        [5] = "Если мне все равно придется ждать, самое меньшее, что ты мог бы сделать — это принести подушку.",
        [6] = "Вовремя. Есть то, что мне нужно?",
        [7] = "Готов выложить товары?",
        [8] = "Неужто это мастер? Я все понимаю — хорошая работа требует времени.",
        [9] = "Мне нравится думать об этом, как о «косвенном покровительстве». Что ж, давай посмотрим, что ты принес своему покровителю.",
        [10] = "Пришел доставить наш маленький 'частный заказ'?",
        [11] = "Большинство клиентов не любит ждать. Вредит репутации, знаешь ли.",
        [12] = "Кажется, ты готов впечатлять.",
        [13] = "Приятно видеть законченную работу. Молодец.",
        [14] = "Замечательно. Я передам это в нужные руки.",
        [15] = "Может быть, в этот раз клиент будет доволен.",
        [16] = "Я удостоверюсь, что это попадет куда надо. Вот твой гонорар.",
        [17] = "Это подойдет. Да, прекрасно подойдет.",
    },
    ["it"] =
    {
        [1] = "This will do. Yes, this will do nicely.",
        [2] = "I'll make sure this goes where it needs to. Here's your fee.",
        [3] = "Excellent. I'll get this to the right hands.",
        [4] = "Good to see the finished product. Nice work.",
        [5] = "Maybe this time the client will be satisfied.",
        [6] = "None of us is getting any younger, you know.",
    },
    ["ja"] =
    {
        [1] = "俺たちは誰も、これ以上若くなったりはしないからな",
        [2] = "形と機能、どっちを重視する？ まあいいさ。とにかく渡してくれ",
        [3] = "悪くないじゃないか。立派なきらめきだ。見事だし、力強い。顧客も喜ぶだろう",
        [4] = "よくやったな。どうやってこんな風に金属を曲げるんだ？ まあいい。これが報酬だ",
        [5] = "完成品が見れて嬉しいよ。良くやってくれた",
        [6] = "必要な者の手に必ず渡すよ。これが報酬だ",
        [7] = "多分、今度は顧客にも満足してもらえるだろう",
        [8] = "やるじゃないか。ふさわしい者の手に届けよう",
        [9] = "これで十分だ。正直に言えば、十二分ってヤツだな",
        [10] = "長く待たせれば、顧客はそれだけ容赦しなくなる。言ってる意味はわかるよな？",
        [11] = "感心させようって顔つきだな",
        [12] = "顧客の多くは待たされるのを好まない。評判が下がるからな",
        [13] = "俺にはいくらでも時間がある。だが、顧客はそうじゃない",
        [14] = "時間だな。必要なものは揃ってるか？",
        [15] = "職人のお出ましか。優れた仕事には時間がかかるって考えることにしてるよ",
        [16] = "俺たちの「個人的契約」を満たしに来たのかな？",
        [17] = "どうせ待たなくちゃならないんだったら、気を利かせて枕ぐらい持ってきてくれよ",
        [18] = "品物を届ける準備はいいか？",
        [19] = "職人が来たな。調子はどうだ？",
        [20] = "俺はこいつを「間接的支援」だと思ってる。俺の後援者のためにあんたが何を持ってきたのか、確かめようじゃないか",
    },
    ["es"] =
    {
        [1] = "Not half bad. Respectable gleam. Nice, strong style. The client will be pleased.",
        [2] = "Nice. How do you bend the metal like that? Anyway, here's your fee.",
        [3] = "Form and function, where do you stand? Oh, never mind. Just hand it over.",
        [4] = "Ninguno de nosotros se hará más joven, ¿sabes?",
        [5] = "Que bueno ver el producto final. Buen trabajo.",
        [6] = "Excelente. Esto llegará a buenas manos.",
        [7] = "Quizá el cliente esté satisfecho esta vez.",
        [8] = "I'll make sure this goes where it needs to. Here's your fee.",
        [9] = "Con esto basta. Sí, es bastante bueno.",
        [10] = "Me gusta pensar en esto como un \"empleo indirecto\". Veamos que tienes para nuestro patrón.",
        [11] = "¿Estás listo para dejar la mercancía?",
        [12] = "¿Vienes a entregar nuestro pequeño \"contrato privado\"?",
        [13] = "Anda, si es el artesano. Tengo entendido que el buen trabajo requiere de tiempo.",
        [14] = "Ha llegado el artesano. ¿Qué tal el trabajo?",
    },
    ["en"] =
    {
        [1] = "Good to see the finished product. Nice work.",
        [2] = "This will do. Yes, this will do nicely.",
        [3] = "Excellent. I'll get this to the right hands.",
        [4] = "Maybe this time the client will be satisfied.",
        [5] = "Good to see the finished product. Nice work.",
        [6] = "Maybe this time the client will be satisfied.",
        [7] = "I'll make sure this goes where it needs to. Here's your fee.",
        [8] = "Excellent. I'll get this to the right hands.",
        [9] = "This will do. Yes, this will do nicely.",
        [10] = "Good to see the finished product. Nice work.",
        [11] = "This will do. Yes, this will do nicely.",
        [12] = "Excellent. I'll get this to the right hands.",
        [13] = "I'll make sure this goes where it needs to. Here's your fee.",
        [14] = "Maybe this time the client will be satisfied.",
        [15] = "This will do. Yes, this will do nicely.",
        [16] = "Maybe this time the client will be satisfied.",
        [17] = "I'll make sure this goes where it needs to. Here's your fee.",
        [18] = "Good to see the finished product. Nice work.",
        [19] = "If I have to wait around anyway, the least you could do is bring me a pillow.",
        [20] = "You look ready to impress.",
        [21] = "I like to think of this as 'indirect patronage'. So, let's see what you have for our patron.",
        [22] = "Most clients don't like to be kept waiting. Hurts a reputation, you know.",
        [23] = "The longer clients wait, the less patient they become. If you know what I mean.",
        [24] = "About time. Got what I need?",
        [25] = "You ready to drop the goods off?",
        [26] = "None of us is getting any younger, you know.",
        [27] = "The artisan arrives. How's work?",
        [28] = "I've got all the time in the world. The client doesn't, of course.",
        [29] = "If it isn't the artisan. I'm given to understand that fine work takes time.",
        [30] = "Come to deliver our little 'private contract'?",
    },
},
